core/src/streams.h

```diff
@@ -0,0 +1,352 @@
+#ifndef _G3_STREAMS_H
+#define _G3_STREAMS_H
+
+#include <G3Logging.h>
+
+#ifdef ZLIB_FOUND
+#include <zlib.h>
+#endif
+#ifdef BZIP2_FOUND
+#include <bzlib.h>
+#endif
+#ifdef LZMA_FOUND
+#include <lzma.h>
+#endif
+
+#include <unistd.h>
+#include <fstream>
+#include <streambuf>
+
+class RemoteInputStreamBuffer : public std::streambuf {
+public:
+	RemoteInputStreamBuffer(int fd, size_t size)
+	  : fd_(fd), buffer_(size), bytes_(0) {
+		setg(buffer_.data(), buffer_.data(), buffer_.data());
+	}
+
+	~RemoteInputStreamBuffer() {
+		close(fd_);
+	}
+
+	int fd() { return fd_; }
+
+protected:
+	int_type underflow() {
+		if (gptr() < egptr())
+			return traits_type::to_int_type(*gptr());
+
+		ssize_t n = read(fd_, buffer_.data(), buffer_.size());
+		if (n <= 0)
+			return traits_type::eof();
+		setg(buffer_.data(), buffer_.data(), buffer_.data() + n);
+		return traits_type::to_int_type(*gptr());
+	}
+
+	std::streamsize xsgetn(char* s, std::streamsize n) {
+		std::streamsize n_read = 0;
+		while (n_read < n) {
+			if (gptr() == egptr()) {
+				if (underflow() == traits_type::eof())
+					break;
+			}
+
+			std::streamsize remaining = n - n_read;
+			std::streamsize available = egptr() - gptr();
+			std::streamsize to_read = std::min(remaining, available);
+
+			std::memcpy(s + n_read, gptr(), to_read);
+			gbump(to_read);
+			n_read += to_read;
+		}
+		bytes_ += n_read;
+		return n_read;
+	}
+
+	std::streampos seekoff(std::streamoff off, std::ios_base::seekdir way,
+	    std::ios_base::openmode mode) {
+		// short-circuit for tellg
+		if ((mode & std::ios_base::in) && off == 0 && way == std::ios_base::cur)
+			return bytes_;
+		log_fatal("Seek not implemented for remote stream");
+	}
+
+	std::streampos seekpos(std::streampos pos, std::ios_base::openmode mode) {
+		log_fatal("Seek not implemented for remote stream");
+	}
+
+private:
+	int fd_;
+	std::vector<char> buffer_;
```

Comment:
The various `*Buffer` classes never seem to resize their `buffer_`s or do any other high-level operations on them, so it seems like `std::unique_ptr<char[]>` would be a more natural type than `std::vector<char>`. (This would require explicitly keeping a variable for `size`, but this would still be a small savings compared to the `vector` keeping both a size and a capacity around.)